src/pykeen/models/nbase.py

```diff
@@ -0,0 +1,475 @@
+        )
+        self.relation_representations = _prepare_representation_module_list(
+            representations=relation_representations,
+            num_embeddings=triples_factory.num_relations,
+            shapes=interaction.relation_shape,
+            label="relation",
+        )
+        self.interaction = interaction
+        # Comment: it is important that the regularizers are stored in a module list, in order to appear in
+        # model.modules(). Thereby, we can collect them automatically.
+        self.weight_regularizers = nn.ModuleList()
+
+    def append_weight_regularizer(
+        self,
+        parameter: Union[str, nn.Parameter, Iterable[Union[str, nn.Parameter]]],
+        regularizer: Regularizer,
+    ) -> None:
+        """Add a model weight to a regularizer's weight list, and register the regularizer with the model.
+
+        :param parameter:
+            The parameter, either as name, or as nn.Parameter object. A list of available parameter names is shown by
+             `sorted(dict(self.named_parameters()).keys())`.
+        :param regularizer:
+            The regularizer instance which will regularize the weights.
+        """
+        # normalize input
+        if isinstance(parameter, (str, nn.Parameter)):
+            parameter = [parameter]
+        weights: Mapping[str, nn.Parameter] = dict(self.named_parameters())
+        for param in parameter:
+            if isinstance(param, str):
+                if param not in weights:
+                    raise KeyError(f"Invalid parameter_name={parameter}. Available are: {sorted(weights.keys())}.")
+                param: nn.Parameter = weights[param]  # type: ignore
+            regularizer.add_parameter(parameter=param)
+        self.weight_regularizers.append(regularizer)
+
+    def forward(
+        self,
+        h_indices: Optional[torch.LongTensor],
+        r_indices: Optional[torch.LongTensor],
+        t_indices: Optional[torch.LongTensor],
+        slice_size: Optional[int] = None,
+        slice_dim: Optional[str] = None,
+    ) -> torch.FloatTensor:
+        """Forward pass.
+
+        This method takes head, relation and tail indices and calculates the corresponding score.
+
+        All indices which are not None, have to be either 1-element or have the same shape, which is the batch size.
+
+        :param h_indices:
+            The head indices. None indicates to use all.
+        :param r_indices:
+            The relation indices. None indicates to use all.
+        :param t_indices:
+            The tail indices. None indicates to use all.
+        :param slice_size:
+            The slice size.
+        :param slice_dim:
+            The dimension along which to slice. From {"h", "r", "t"}
+
+        :return: shape: (batch_size, num_heads, num_relations, num_tails)
+            The score for each triple.
+        """
+        h, r, t = self._get_representations(h_indices, r_indices, t_indices)
```

Comment:
Pass by keyword instead of positional? Makes it less prone to errors when changing parameter order.

Reply:
Done in 50245c60